lib/PrometheusPlugin.ts

```diff
@@ -23,81 +23,157 @@
+          gcDurationBuckets: [0.001, 0.01, 0.1, 1, 2, 5],
+        },
+        core: {
+          monitorRequestDuration: true,
+          prefix: 'kuzzle_',
+        }
       }
     };
   }
 
-  async init (config: JSONObject, context: PluginContext) {
-    this.config = { ...this.defaultConfig, ...config };
+  /**
+   * Plugin initialization
+   * @param {PrometheusPluginConfiguration} config  - Plugin configuration
+   * @param {PluginContext}       context - Kuzzle plugin context
+   */
+  async init(config: PrometheusPluginConfiguration, context: PluginContext) {
+    this.config = _.merge(this.config, config);
     this.context = context;
 
-    this.prometheusController = new PrometheusController(this.config, this.context);
-    await this.prometheusController.init();
-
-    this.api = {
-      'prometheus': this.prometheusController.definition
+    this.pipes = {
+      'server:afterMetrics': async (request: KuzzleRequest) => this.pipeFormatMetrics(request),
     };
 
     this.hooks = {
-      'connection:new': () => this.prometheusController.recordConnections('inc'),
-      'connection:remove': () => this.prometheusController.recordConnections('dec'),
-      'core:realtime:room:create:after': () => this.prometheusController.recordRooms('inc'),
-      'core:realtime:room:remove:before': () => this.prometheusController.recordRooms('dec'),
-      'core:realtime:user:subscribe:after': () => this.prometheusController.recordSubscriptions('inc'),
-      'core:realtime:user:unsubscribe:after': () => this.prometheusController.recordSubscriptions('dec'),
-      'request:onSuccess': (request, event) => this.prometheusController.recordRequests(request, event),
-      'request:onError': (request, event) => this.prometheusController.recordRequests(request, event)
+      'request:on*': this.recordRequest.bind(this),
+      //'request:onError': (request, event) => this.prometheusController.recordRequests(request, event)
     };
+
+    this.metricService = new MetricService(this.config as PrometheusPluginConfiguration);
+  }
+
+  /**
+   * On server:afterMetrics hook, format the metrics and send them to the client as Prometheus format
+   * @param {KuzzleRequest} request - Kuzzle request
+   * @returns {KuzzleRequest}
+   */
+  async pipeFormatMetrics (request: KuzzleRequest): Promise<KuzzleRequest> {
+    // coreMetrics need to be updated with Kuzzle core values before the metrics are sent to the client
+    this.metricService.updateCoreMetrics(request.response.result);
+
+    request.response.configure({
+      headers: {
+        'Content-Type': this.metricService.getPrometheusContentType()
+      },
+      format: 'raw',
+    });
+    request.response.result = await this.metricService.getMetrics();
+
+    return request;
+  }
+
+  /**
+   * Log the response time for the given request in the associated metric
+   * @param {KuzzleRequest} request - Kuzzle request
+   */
+  recordRequest(request: KuzzleRequest): void {
```

Comment:
Nitpicking: coding standard
```suggestion
  recordRequest (request: KuzzleRequest): void {
```